### ISSUE TITLE
[refactor] improve QEOM code

### ISSUE THREAD

**Issue by mrossinek** [Qiskit/qiskit-nature#769]
### What is the expected enhancement?

The implementation of the `QEOM` algorithm looks like it could undergo some refactoring in order to make the code more maintainable. I do not have any specific targets in mind right now, but I am opening this issue to keep this in mind for some point in the future.

Some rough ideas include:
- improve the hopping operator generation and commutativity handling:
    - right now, this is split across 5+ methods which can surely be improved to some degree
    - the code needs to handle tapering manually. This results in a very hard to follow flow of logic. This can surely be streamlined quite a bit better
    - `parallel_map` is being used in multiple places. We should check if we can streamline this better in order to avoid multiple iterations over the same data
- generally add more detailed/explanatory comments in the code

**Comment by mrossinek** [comment]
@Anthony-Gandon would you be interested in working on this, given your recent review of #768?

**Comment by Anthony-Gandon** [comment]
Yes you can assign this to me. Thanks!
